kloppy/domain/services/state_builder/base.py

```diff
@@ -0,0 +1,16 @@
+from abc import abstractmethod, ABC
+from typing import TypeVar
+
+from kloppy.domain import EventDataset, Event
+
+T = TypeVar("T")
+
+
+class StateBuilder(ABC):
+    @abstractmethod
+    def initial_state(self, dataset: EventDataset) -> T:
+        pass
+
+    @abstractmethod
+    def reduce(self, state: T, event: Event) -> T:
```

Comment:
This signature makes it hard to do lookahead and lookbehind for more interesting types of state. You could hide things in state to communicate through time if you wanted to but that seems messy. You also can't build a state builder that itself relies on state from another builder.

I'd much rather see something like Clojure's transducers here - they give you the option of maintaining internal state across elements of the collection, but can also compose together to create actual pipelines.

Reply:
I'll move this to the next version.